File: 153_Find Minimum in Rotated Sorted Array.java
// Time complexity: O(log(n))
// Space complexity: O(1)

class Solution {
    public int findMin(int[] nums) {
        // label [start, ..., a[n-1],] as RED and [a[0], ..., end] as BLUE
        // the goal is to find the first BLUE

        int left = 0, right = nums.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            // [mid, right] is BLUE
            if (nums[mid] < nums[right]) right = mid;
            // [mid, right] is RED + BLUE && nums[mid] is RED
            else left = mid + 1;
        }

        return nums[right];
    }
}

File: 153_Find Minimum in Rotated Sorted Array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # [greater, smaller]
        l, r = 0, len(nums) - 1

        while l < r:
            # 左中点会偏左, 保证r=m时更新还会动 (r=m w/ 左中点 and l=m w/右中点)
            m = (l + r) >> 1
            # case: in the greater area -> [m+1, r]
            if nums[m] > nums[r]: l = m + 1
            # case: in the smaller area -> [l, m]
            else: r = m
        
        return nums[l]